File: src/main/java/io/github/saneea/feature/binary/ToHex.java
package io.github.saneea.feature.binary;

import java.io.IOException;
import java.io.InputStream;
import java.io.PrintStream;

import io.github.saneea.Feature;
import io.github.saneea.FeatureContext;

public class ToHex implements Feature, Feature.In.Bin.Stream, Feature.Out.Text.PrintStream {

	private InputStream in;
	private PrintStream out;

	@Override
	public Meta meta(FeatureContext context) {
		return Meta.from("convert input binary sequence to hex");
	}

	@Override
	public void run(FeatureContext context) throws IOException {
		run(in, out);
	}

	public static void run(InputStream input, PrintStream out) throws IOException {
		int charCode;
		while ((charCode = input.read()) != -1) {

			String hexString = Integer.toHexString(charCode);
			if (hexString.length() < 2) {
				out.print('0');
			}
			out.print(hexString);
		}
	}

	@Override
	public void setIn(InputStream in) {
		this.in = in;
	}

	@Override
	public void setOut(PrintStream out) {
		this.out = out;
	}

}

File: src/main/kotlin/io/github/saneea/dvh/feature/binary/hex/ToHex.kt
package io.github.saneea.dvh.feature.binary.hex

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import java.io.InputStream
import java.io.PrintStream

class ToHex :
    Feature,
    Feature.In.Bin.Stream,
    Feature.Out.Text.PrintStream {

    override lateinit var context: FeatureContext
    override lateinit var inBinStream: InputStream
    override lateinit var outTextPrintStream: PrintStream

    override val meta get() = Meta("convert input binary sequence to hex")

    override fun run() {
        run(inBinStream, outTextPrintStream)
    }

    companion object {
        fun run(input: InputStream, out: PrintStream) {
            var charCode: Int
            while (input.read().also { charCode = it } != -1) {
                val hexString = Integer.toHexString(charCode)
                if (hexString.length < 2) {
                    out.print('0')
                }
                out.print(hexString)
            }
        }
    }
}